packages/puppeteer-core/src/api/Frame.ts

```diff
@@ -720,13 +721,12 @@ export abstract class Frame extends EventEmitter<FrameEvents> {
     selector: Selector,
     options: WaitForSelectorOptions = {}
   ): Promise<ElementHandle<NodeFor<Selector>> | null> {
-    const {updatedSelector, QueryHandler} =
+    const {updatedSelector, QueryHandler, selectorHasPseudoClasses} =
       getQueryHandlerAndSelector(selector);
-    return (await QueryHandler.waitFor(
-      this,
-      updatedSelector,
-      options
-    )) as ElementHandle<NodeFor<Selector>> | null;
+    return (await QueryHandler.waitFor(this, updatedSelector, {
+      ...options,
+      polling: selectorHasPseudoClasses ? PollingOptions.RAF : undefined,
```

Comment:
We need change the order else the `polling` from the user is not respected.
Or we need instead of `undefined` -> `options.polling`
```suggestion
      polling: selectorHasPseudoClasses ? PollingOptions.RAF : undefined,
      ...options,
```
![image](https://github.com/puppeteer/puppeteer/assets/34244704/7cf3acb7-9909-4621-9595-211e76c6879a)

Reply:
Wait but WaitForSelectorOptions does not allow polling?